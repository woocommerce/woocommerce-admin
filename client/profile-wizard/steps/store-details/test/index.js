/**
 * External dependencies
 */
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import { StoreDetails } from '../';

const testProps = {
	query: {
		page: 'wc-admin',
		path: '/setup-wizard',
	},
	step: {
		key: 'store-details',
		label: 'Store Details',
		isComplete: false,
	},
	initialValues: {
		addressLine1: '',
		addressLine2: '',
		city: '',
		countryState: '',
		postCode: '',
		isAgreeMarketing: true,
		storeEmail: 'wordpress@example.com',
	},
	getLocale: jest.fn(),
	isLoading: false,
};

describe( 'StoreDetails', () => {
	describe( 'Snapshot test', () => {
		test( 'should match saved snapshot', () => {
			const container = render( <StoreDetails { ...testProps } /> );
			expect( container ).toMatchSnapshot();
		} );
	} );
	describe( 'Email validation test cases', () => {
		// test cases taken from wordpress php is_email test cases
		// https://github.com/WordPress/wordpress-develop/blob/2648a5f984b8abf06872151898e3a61d3458a628/tests/phpunit/tests/formatting/isEmail.php
		test.each( [
			'khaaaaaaaaaaaaaaan!',
			'http://bob.example.com/',
			"sif i'd give u it, spamer!1",
			'com.exampleNOSPAMbob',
			'bob@your mom',
			'a@b.c',
		] )( 'should fail email validation when given %s', async ( email ) => {
			const container = render( <StoreDetails { ...testProps } /> );
			const emailInput = container.getByLabelText( 'Email address' );
			await userEvent.clear( emailInput );
			await userEvent.type( emailInput, email );
			// validation is triggered onChange but error message only renders on blur
			// react testing lib doesn't have a "blur" event that can be triggered so this does the job of triggering the error message rendering
			userEvent.tab();
			expect(
				container.queryByText( 'Invalid email address' )
			).toBeInTheDocument();
		} );

		test.each( [
			'bob@example.com',
			'phil@example.info',
			// 'ace@204.32.222.14', this testcase passes for the backend validation but fails here, following up in a PR to fix this in @wordpress/url
			'kevin@many.subdomains.make.a.happy.man.edu',
			'a@b.co',
			'bill+ted@example.com',
		] )( 'should pass email validation when given %s', async ( email ) => {
			const container = render( <StoreDetails { ...testProps } /> );
			const emailInput = container.getByLabelText( 'Email address' );
			await userEvent.clear( emailInput );
			await userEvent.type( emailInput, email );
			userEvent.tab();
			expect(
				container.queryByText( 'Invalid email address' )
			).toBeNull();
		} );

		test( 'should pass email validation when field is empty', async () => {
			const container = render( <StoreDetails { ...testProps } /> );
			const emailInput = container.getByLabelText( 'Email address' );
			await userEvent.clear( emailInput );
			userEvent.tab();
			expect(
				container.queryByText( 'Invalid email address' )
			).toBeNull();
		} );
	} );
} );
