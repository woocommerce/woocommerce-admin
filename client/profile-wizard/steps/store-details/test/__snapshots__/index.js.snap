// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreDetails Snapshot test should match saved snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <p
      class="a11y-speak-intro-text"
      hidden="hidden"
      id="a11y-speak-intro-text"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    >
      Notifications
    </p>
    <div
      aria-atomic="true"
      aria-live="assertive"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-assertive"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-polite"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div>
      <div
        class="woocommerce-profile-wizard__store-details"
      >
        <div
          class="woocommerce-profile-wizard__step-header"
        >
          <h2
            class="css-1ahfdc3-Text e15wbhsk0"
            size="20"
          >
            Welcome to WooCommerce
          </h2>
          <p
            class="css-1f0yw52-Text e15wbhsk0"
          >
            Tell us about your store and we'll get you set up in no time
            <button
              aria-label="Learn more about store details"
              class="components-button is-tertiary"
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z"
                />
              </svg>
            </button>
          </p>
        </div>
        <div
          class="components-card is-size-medium css-1xs3c37-CardUI e1q7k77g0"
        >
          <div
            class="components-card__body is-size-medium css-xmjzce-BodyUI e1q7k77g3"
          >
            <span
              class="components-spinner"
            />
            <div>
              <div
                class="components-base-control muriel-component muriel-input-text with-value"
              >
                <div
                  class="components-base-control__field"
                >
                  <label
                    class="components-base-control__label"
                    for="inspector-text-control-0"
                  >
                    Email address
                  </label>
                  <input
                    autocomplete="email"
                    checked=""
                    class="components-text-control__input"
                    id="inspector-text-control-0"
                    placeholder="Email address"
                    required=""
                    type="text"
                    value="wordpress@example.com"
                  />
                </div>
              </div>
            </div>
            <div
              class="components-flex__item css-1s295sp-Item eboqfv51"
            >
              <div
                class="woocommerce-profile-wizard__newsletter-signup"
              >
                <div
                  class="components-base-control css-wdf2ti-Wrapper e1puf3u0"
                >
                  <div
                    class="components-base-control__field css-11vcxb9-StyledField e1puf3u1"
                  >
                    <span
                      class="components-checkbox-control__input-container"
                    >
                      <input
                        checked=""
                        class="components-checkbox-control__input"
                        id="inspector-checkbox-control-0"
                        type="checkbox"
                        value="true"
                      />
                      <svg
                        aria-hidden="true"
                        class="components-checkbox-control__checked"
                        focusable="false"
                        height="24"
                        role="img"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M18.3 5.6L9.9 16.9l-4.6-3.4-.9 1.2 5.8 4.3 9.3-12.6z"
                        />
                      </svg>
                    </span>
                    <label
                      class="components-checkbox-control__label"
                      for="inspector-checkbox-control-0"
                    >
                      Get tips, product updates and inspiration straight to your mailbox.
                       
                      <span
                        class="woocommerce-profile-wizard__powered-by-mailchimp"
                      >
                        Powered by Mailchimp
                      </span>
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="components-flex components-card__footer is-size-medium e1q7k77g4 css-1fpr7ir-Flex-FooterUI eboqfv50"
          >
            <button
              class="components-button is-primary"
              disabled=""
              type="button"
            >
              Continue
            </button>
          </div>
        </div>
        <div
          class="woocommerce-profile-wizard__footer"
        >
          <button
            class="components-button woocommerce-profile-wizard__footer-link is-link"
            type="button"
          >
            Skip setup store details
          </button>
          <button
            aria-label="Manual setup is only recommended for
 experienced WooCommerce users or developers."
            class="components-button is-tertiary"
            type="button"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="woocommerce-profile-wizard__store-details"
    >
      <div
        class="woocommerce-profile-wizard__step-header"
      >
        <h2
          class="css-1ahfdc3-Text e15wbhsk0"
          size="20"
        >
          Welcome to WooCommerce
        </h2>
        <p
          class="css-1f0yw52-Text e15wbhsk0"
        >
          Tell us about your store and we'll get you set up in no time
          <button
            aria-label="Learn more about store details"
            class="components-button is-tertiary"
            type="button"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z"
              />
            </svg>
          </button>
        </p>
      </div>
      <div
        class="components-card is-size-medium css-1xs3c37-CardUI e1q7k77g0"
      >
        <div
          class="components-card__body is-size-medium css-xmjzce-BodyUI e1q7k77g3"
        >
          <span
            class="components-spinner"
          />
          <div>
            <div
              class="components-base-control muriel-component muriel-input-text with-value"
            >
              <div
                class="components-base-control__field"
              >
                <label
                  class="components-base-control__label"
                  for="inspector-text-control-0"
                >
                  Email address
                </label>
                <input
                  autocomplete="email"
                  checked=""
                  class="components-text-control__input"
                  id="inspector-text-control-0"
                  placeholder="Email address"
                  required=""
                  type="text"
                  value="wordpress@example.com"
                />
              </div>
            </div>
          </div>
          <div
            class="components-flex__item css-1s295sp-Item eboqfv51"
          >
            <div
              class="woocommerce-profile-wizard__newsletter-signup"
            >
              <div
                class="components-base-control css-wdf2ti-Wrapper e1puf3u0"
              >
                <div
                  class="components-base-control__field css-11vcxb9-StyledField e1puf3u1"
                >
                  <span
                    class="components-checkbox-control__input-container"
                  >
                    <input
                      checked=""
                      class="components-checkbox-control__input"
                      id="inspector-checkbox-control-0"
                      type="checkbox"
                      value="true"
                    />
                    <svg
                      aria-hidden="true"
                      class="components-checkbox-control__checked"
                      focusable="false"
                      height="24"
                      role="img"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M18.3 5.6L9.9 16.9l-4.6-3.4-.9 1.2 5.8 4.3 9.3-12.6z"
                      />
                    </svg>
                  </span>
                  <label
                    class="components-checkbox-control__label"
                    for="inspector-checkbox-control-0"
                  >
                    Get tips, product updates and inspiration straight to your mailbox.
                     
                    <span
                      class="woocommerce-profile-wizard__powered-by-mailchimp"
                    >
                      Powered by Mailchimp
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="components-flex components-card__footer is-size-medium e1q7k77g4 css-1fpr7ir-Flex-FooterUI eboqfv50"
        >
          <button
            class="components-button is-primary"
            disabled=""
            type="button"
          >
            Continue
          </button>
        </div>
      </div>
      <div
        class="woocommerce-profile-wizard__footer"
      >
        <button
          class="components-button woocommerce-profile-wizard__footer-link is-link"
          type="button"
        >
          Skip setup store details
        </button>
        <button
          aria-label="Manual setup is only recommended for
 experienced WooCommerce users or developers."
          class="components-button is-tertiary"
          type="button"
        >
          <svg
            aria-hidden="true"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
